### PR TITLE
added placement group as an option for autoscaling groups

### PR DESCRIPTION
**Pull request by dayer4b**
in reference to issue #3703

### CODE REVIEW

File: builtin/providers/aws/resource_aws_autoscaling_group.go [hunk 1]
```diff
@@ -95,6 +95,13 @@ func resourceAwsAutoscalingGroup() *schema.Resource {
 				Set:      schema.HashString,
 			},
 
+			"placement_group": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_autoscaling_group.go]: I just think the addition to the schema should be really simple:

TypeString
Optional

This goes as per the fact the [docs](http://docs.aws.amazon.com/cli/latest/reference/autoscaling/create-auto-scaling-group.html) says that the input is a string and can be empty

File: builtin/providers/aws/resource_aws_autoscaling_group.go [hunk 2]
```diff
@@ -175,6 +182,11 @@ func resourceAwsAutoscalingGroupCreate(d *schema.ResourceData, meta interface{})
 		autoScalingGroupOpts.HealthCheckGracePeriod = aws.Int64(int64(v.(int)))
 	}
 
+	if v, ok := d.GetOk("placement_group"); ok && v.(*schema.Set).Len() > 0 {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_autoscaling_group.go]: USing availability_zones is not quite right for the placement_group. The `health_check_type` is a closer match to the system

**Comment by stack72** [comment]
@dayer4b you have only catered for the Create method. You need to make sure that placement_group is also taken care of in the Update method as well

**Comment by dayer4b** [comment]
@stack72 thanks for all the help! I've caught up with your suggestions and updated my branch.

**Comment by stack72** [comment]
@dayer4b, this looks much better

I just ran your tests and got the following:

```=== RUN   TestAccAWSAutoScalingGroup_basic
--- FAIL: TestAccAWSAutoScalingGroup_basic (7.20s)
	testing.go:137: Step 0 error: Error applying: 1 error(s) occurred:

		* aws_autoscaling_group.bar: Error creating Autoscaling Group: ValidationError: Placement group test is invalid: The Placement Group 'test' is unknown.
			status code: 400, request id: 7403f6e8-828b-11e5-93c6-b9e2a4ee28a3
```

In order to fix these (verified), you need to make the test config as follows:

```
const testAccAWSAutoScalingGroupConfig = `
resource "aws_launch_configuration" "foobar" {
  image_id = "ami-21f78e11"
  instance_type = "t1.micro"
}

resource "aws_placement_group" "test" {
  name = "test"
  strategy = "cluster"
}

resource "aws_autoscaling_group" "bar" {
  availability_zones = ["us-west-2a"]
  name = "foobar3-terraform-test"
  max_size = 5
  min_size = 2
  health_check_grace_period = 300
  health_check_type = "ELB"
  desired_capacity = 4
  force_delete = true
  termination_policies = ["OldestInstance","ClosestToNextInstanceHour"]
  placement_group = "${aws_placement_group.test.id}"

  launch_configuration = "${aws_launch_configuration.foobar.name}"

  tag {
    key = "Foo"
    value = "foo-bar"
    propagate_at_launch = true
  }
}
`
```

Notice that the test now sets up `resource "aws_placement_group" "test"`

Paul

**Comment by dayer4b** [comment]
oh, of course.  that makes a lot of sense.  sorry. I'll get on that

**Comment by dayer4b** [comment]
Just an update - I added the `aws_placement_group` resource last week.  Sorry that I didn't make that clear.

**Comment by stack72** [comment]
thanks @dayer4b - I will give your tests a shot again :)

**Comment by stack72** [comment]
@catsby I have run the tests here and they look ok. I think this is worth a final review by you or @phinze :)

**Comment by stack72** [comment]
@dayer4b ok, I re-pulled your PR and the basic test actually doesn't work. 

Currently, your test config looks as follows:

```
resource "aws_launch_configuration" "foobar" {
  image_id = "ami-21f78e11"
  instance_type = "t1.micro"
}

resource "aws_autoscaling_group" "bar" {
  availability_zones = ["us-west-2a"]
  name = "foobar3-terraform-test"
  max_size = 5
  min_size = 2
  health_check_grace_period = 300
  health_check_type = "ELB"
  desired_capacity = 4
  force_delete = true
  termination_policies = ["OldestInstance","ClosestToNextInstanceHour"]
  placement_group = "test"

  launch_configuration = "${aws_launch_configuration.foobar.name}"

  tag {
    key = "Foo"
    value = "foo-bar"
    propagate_at_launch = true
  }
}
```

This results in a test error as follows:

```
=== RUN   TestAccAWSAutoScalingGroup_basic
--- FAIL: TestAccAWSAutoScalingGroup_basic (6.14s)
	testing.go:137: Step 0 error: Error applying: 1 error(s) occurred:

		* aws_autoscaling_group.bar: Error creating Autoscaling Group: ValidationError: Placement group test is invalid: The Placement Group 'test' is unknown.
			status code: 400, request id: 60ebc73c-8894-11e5-ae11-fd4e0662e4f6
```

Your config needs to look as follows:

```
resource "aws_launch_configuration" "foobar" {
  image_id = "ami-21f78e11"
  instance_type = "t1.micro"
}

resource "aws_placement_group" "test" {
  name = "test"
  strategy = "cluster"
}

resource "aws_autoscaling_group" "bar" {
  availability_zones = ["us-west-2a"]
  name = "foobar3-terraform-test"
  max_size = 5
  min_size = 2
  health_check_grace_period = 300
  health_check_type = "ELB"
  desired_capacity = 4
  force_delete = true
  termination_policies = ["OldestInstance","ClosestToNextInstanceHour"]
  placement_group = "${aws_placement_group.test.id}"

  launch_configuration = "${aws_launch_configuration.foobar.name}"

  tag {
    key = "Foo"
    value = "foo-bar"
    propagate_at_launch = true
  }
}

```

You only added the placement_group resource to the documentation

**Comment by dayer4b** [comment]
wow, i had made the change and left it unstaged.  sorry.  i just pushed.

**Comment by dayer4b** [comment]
are there any remaining issues with this?  I can amend my branch if need be

**Comment by stack72** [comment]
@dayer4b ok, so I've just re-tested this and have some final recommendations (I promise!) - recommendations inline

File: builtin/providers/aws/resource_aws_autoscaling_group_test.go [hunk 3]
```diff
@@ -364,6 +371,7 @@ resource "aws_autoscaling_group" "bar" {
   desired_capacity = 4
   force_delete = true
   termination_policies = ["OldestInstance","ClosestToNextInstanceHour"]
+  placement_group = "${aws_placement_group.test.name}"
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_autoscaling_group_test.go]: This needs to come out of the _basic test

File: builtin/providers/aws/resource_aws_autoscaling_group_test.go [hunk 4]
```diff
@@ -354,6 +356,11 @@ resource "aws_launch_configuration" "foobar" {
   instance_type = "t1.micro"
 }
 
+resource "aws_placement_group" "test" {
+  name = "test"
+  strategy = "cluster"
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_autoscaling_group_test.go]: This needs to come out of the _basic test

File: builtin/providers/aws/resource_aws_autoscaling_group_test.go [hunk 5]
```diff
@@ -48,6 +48,8 @@ func TestAccAWSAutoScalingGroup_basic(t *testing.T) {
 						"aws_autoscaling_group.bar", "termination_policies.0", "OldestInstance"),
 					resource.TestCheckResourceAttr(
 						"aws_autoscaling_group.bar", "termination_policies.1", "ClosestToNextInstanceHour"),
+					resource.TestCheckResourceAttr(
+						"aws_autoscaling_group.bar", "placement_group", "test"),
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_autoscaling_group_test.go]: This needs to come out of the _basic test

**Comment by phinze** [comment]
Will merge this and make @stack72's recommended tweaks on master. :+1: 

Thanks for your work here, @dayer4b!

**Pull request merged**